convert2rhel/utils.py

```diff
@@ -66,6 +68,196 @@ class Color(object):
+
+        .. important::
+            Exceptions that bubble up to this point will deadlock if the
+            traceback are over 64KiB in size. It is not a good idea to let
+            unhandled exceptions go up this point as it could cause the threads
+            to lock indefinitely.
+            Link to comment: https://stackoverflow.com/questions/19924104/python-multiprocessing-handling-child-errors-in-parent/33599967#comment113491531_33599967
+        """
+        try:
+            multiprocessing.Process.run(self)
+            self._cconn.send(None)
+        # Here, `SystemExit` inherits from `BaseException`, which is too
+        # broad to catch as it involves for-loop exceptions. The idea here
+        # is to catch `SystemExit` *and* any `Exception` that shows up as we do
+        # a lot of logger.critical() and they do raise `SystemExit`.
+        except (Exception, SystemExit) as e:
+            try:
+                import cPickle as pickle
+            except ImportError:
+                import pickle
+
+            try:
+                self._cconn.send(e)
+            except pickle.PicklingError:
+                self._cconn.send(UnableToSerialize("Child process raised %s: %s" % (type(e), str(e))))
+
+    @property
+    def exception(self):
+        if self._pconn.poll():
+            self._exception = self._pconn.recv()
+        return self._exception
+
+
+def run_as_child_process(func):
+    """Decorator to execute a function in a child process.
+
+    This decorator will use `multiprocessing.Process` class to initiate the
+    function as a child process to the parent process with the intention to
+    execute that in its own process, thus, avoiding cases where libraries would
+    install signal handlers that could be propagated to the main thread if they
+    do not do a proper clean-up.
+
+    .. note::
+        This decorator is mostly intended to be used when dealing with
+        functions that call directly `rpm` or `yum`.
+
+        In `rpm` (version 4.11.3, RHEL 7), it's known that whenever there is a
+        call to that library, it will install a global signal handler catching
+        different types of signals, but most importantly, the SIGINT (Ctrl + C)
+        one, which causes a problem during the conversion as we can't replace
+        or override that signal as it was initiated in a C library rather than
+        a python library.
+
+        By using this decorator, `rpm` will install the signal handler in the
+        child process and leave the parent one with the original signal
+        handling that is initiated by python itself (or, whatever signal is
+        registered when the conversion starts as well).
+
+    .. important::
+        If a function that creates child processes is decorated with this
+        decorator, then it won't be possible to use it right away, since this
+        decorator spawn child processes as a daemon (which are normal
```

Comment:
I'm a little unclear on the connection between "it won't be possible to use it right away", "because it spawns child processes as a daemon" and "but can't spawn child processes inside that child process".  Are these three things related?  Or two separate limitations and the reason (daemon processes aren't immediately usable and can't spawn children of their own?)?

Reply:
updated the comment